### DISCUSSION TITLE
Runtime export in webGL environment.

### DISCUSSION THREAD

**Discussion by Tripl3Design** [atteneder/glTFast#455] [General]
As described in the docs, the runtime export can be used to store and share dynamic, user-generated 3D content.
The provided code snippets to save a gltf or glb to a file on your system works fantastic, but is there a possibility to export to a gltf or glb from an online (webgl) environment to the (hosting) server? That would be a really cool way to generate 3D user generated content.

**Comment by atteneder** [comment]
Hi @Tripl3Design,

That's a great use-case indeed! I never tried, but my approach would be to generate a glb into a memory stream and then:

- either upload to a server to host it there
- use some Javascript to make a direct, client-side download/save as file

`GameObjectExport`/`IGltfWritable` provides `Task<bool> SaveToStreamAndDispose(Stream stream)` for that.

Would be great if you could give it a shot and report if you found a working solution. 

Thanks

**Reply by Tripl3Design** [reply]
I can't do something like:
`bool success = await export.SaveToFileAndDispose('https://destination.here/file.glb');`
(hosted on a server of course) right?

A stream is a premade .glb wich constently updates itself?

**Reply by atteneder** [reply]
> I can't do something like: `bool success = await export.SaveToFileAndDispose('https://destination.here/file.glb');` (hosted on a server of course) right?

No, I don't think so. That method is for writing to local files and in WebGL you don't have access to the local file system nor direct access to a server's file system.

> A stream is a premade .glb wich constently updates itself?

A [Stream](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream?view=net-6.0) is an abstract view to a sequence of bytes. You can also write to a stream. A [MemoryStream](https://docs.microsoft.com/en-us/dotnet/api/system.io.memorystream?view=net-6.0) stores the result in the memory. You could use that to generate the glb and in a following step, upload that stream's content to the server via HTTP PUT (you'll need a server-side solution that accepts the upload as well).

There's also a way to [locally save the stream using some Javascript](https://stackoverflow.com/questions/3665115/how-to-create-a-file-in-memory-for-user-to-download-but-not-through-server), but I never tried it myself.

hth

**Reply by Tripl3Design** [reply]
I tried but no succes, am I on the right way?
```
...
  Stream glbExport = new MemoryStream();
  bool success = await export.SaveToStreamAndDispose(glbExport);

  StartCoroutine(GltfExport());
}

IEnumerator SaveGlb()
{
  UnityWebRequest www = UnityWebRequest.Put("https://url.cc/model", glbExport);
  yield return www.SendWebRequest();
....
}
```

Do I have to specify a filepath?

**Reply by atteneder** [reply]
I think you intended to invoke `StartCoroutine(SaveGlb());`, right?
Other than that: I cannot possibly help without knowing what errors you observe.

**Reply by Tripl3Design** [reply]
O yes that is really stupid.

Well I have now:
`async void GltfExport()
    {
        var rootLevelNodes = GameObject.FindGameObjectsWithTag("ExportMe");
        var export = new GameObjectExport();
        export.AddScene(rootLevelNodes);

        Stream glbExport = new MemoryStream();
        bool success = await export.SaveToStreamAndDispose(glbExport);

        Debug.Log("Export = " + glbExport);
        StartCoroutine(SaveGlb(glbExport));

        if (!success)
        {
            Debug.LogError("Something went wrong exporting a glTF");
        }
    }

    IEnumerator SaveGlb(Stream glbExport)
    {
        var memoryStream = new MemoryStream();
        glbExport.CopyTo(memoryStream);

        byte[] byteArray = memoryStream.ToArray();
        UnityWebRequest www = UnityWebRequest.Put("https://etc...", byteArray);
        yield return www.SendWebRequest();

        if (www.result != UnityWebRequest.Result.Success)
        {
            Debug.Log("Error");
        }
        else
        {
            Debug.Log("Upload complete!");
        }
    }`

And the eror in the console is:
`MethodAccessException: Attempt to access method 'GLTFast.ICodeLogger.Error' on type '' failed.
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x00000] in <00000000000000000000000000000000>:0 `


[working demo](https://furnitise.nl/index_rb.php?brand=pastoe&product=wire)

**Reply by atteneder** [reply]
I suspect an error during export. Does it work in the Editor? Because that's way more efficient to iterate. If not, is the error message more verbose if you run a debug build?

You're not waiting for `StartCoroutine(SaveGlb(glbExport));` to finish! Within an `IEnumerator` you could do `yield return StartCoroutine(SaveGlb(glbExport));`, but your GltfExport is an `async void`, so you'd have to wrap it.

no need to copy glbExport to memoryStream. Use glbExport right away.

**Reply by Tripl3Design** [reply]
Is it possible to test a `SaveToStreamAndDispose` in de Editor?
If i do this:
`    async void GltfExport()
    {
        var rootLevelNodes = GameObject.FindGameObjectsWithTag("ExportMe");
        var export = new GameObjectExport();
        export.AddScene(rootLevelNodes);

        MemoryStream memoryStream = new MemoryStream();
        bool success = await export.SaveToStreamAndDispose(memoryStream);

        if (!success)
        {
            Debug.LogError("Something went wrong exporting a glTF");
        }
    }`
I get a `NullReferenceException: Object reference not set to an instance of an object`

**Reply by atteneder** [reply]
I consider figuring out a `NullReferenceException` a basic skill of software development. If you're not up to that task or able to teach it yourself frankly I don't want to spend the time teaching it to you. Good luck anyways.

I might create an real-time export demo some day.

**Reply by Tripl3Design** [reply]
Yes, you are completely right. It's a noob error. Thanks for your time.
SaveToFile works perfectly.
I'm having trouble with SaveToStream. I'm probably not sending the right arguments. Because I don't quite understand that system.IO.Stream.

**Comment by Tripl3Design** [comment]
I did a bit of research, and this works for me:

`async void GltfExport()
    {
        var logger = new CollectingLogger();
        var exportSettings = new ExportSettings
        {
            format = GltfFormat.Binary,
            fileConflictResolution = FileConflictResolution.Overwrite
};
        var export = new GameObjectExport(exportSettings, logger: logger);
        var rootLevelNodes = GameObject.FindGameObjectsWithTag("ExportMe");
        export.AddScene(rootLevelNodes, "My new glTF scene");

        //for local testing
        //var fileName = "model.glb";
        //var outStream = new FileStream(Path.Combine(path, fileName), FileMode.Create);
        var outStream = new MemoryStream();
        bool success = await export.SaveToStreamAndDispose(outStream);

        if (!success)
        {
            Debug.LogError("Something went wrong exporting a glTF");
            logger.LogAll();
        }

        StreamToBytes(outStream);
    }

    public void StreamToBytes(Stream outStream)
    {
        var bytes = new byte[outStream.Length];
        outStream.Seek(0, SeekOrigin.Begin);
        outStream.ReadAsync(bytes, 0, bytes.Length);
        outStream.Dispose();

        StartCoroutine(UploadGlbPost(bytes));
    }

    IEnumerator UploadGlbPost(byte[] bytes)
    {
        List<IMultipartFormSection> formData = new List<IMultipartFormSection>();
        formData.Add(new MultipartFormFileSection("glbfile", bytes, "tempname.glb", "model/gltf-binary"));
        UnityWebRequest www = UnityWebRequest.Post(path, formData);
        yield return www.SendWebRequest();

        if (www.result != UnityWebRequest.Result.Success)
        {
            Debug.Log(www.error);
        }
        else
        {
            Debug.Log("glb upload complete!");
        }
    }`

End then use a little php:

`
$mydata = file_get_contents($_FILES["glbfile"]["tmp_name"]);
$fileName = 'ArModels/frame.glb';
         
$fh = fopen($fileName , 'wb');
fwrite($fh, $mydata);
fclose($fh);
`

Maybe it's possible without php? The PUT request gave a 405 error. Mybe that's a server issue?
I store my textures also on the server and grap them with `UnityWebRequestTexture`. Is there a way to use gltf's instead of glb's and use the online textures?

**Reply by atteneder** [reply]
Great. Glad you figured it all out!

Yes, you can load JSON based .gltf as well.

**Reply by Tripl3Design** [reply]
Well, I need a solution to add the textures (wich are already on the server) to the glb. Or gltf, but only a self-contained gltf is possible when using a stream.

This is the source (Unity webgl):
![source](https://user-images.githubusercontent.com/56125681/187635353-790c03fc-3799-45b9-b536-990e59f61050.jpg)
And this the exported glb without the the textures wich are on the server.
![glb](https://user-images.githubusercontent.com/56125681/187636433-98cafd86-72f9-4591-a3a1-6fbae34a81f9.jpg)

EDIT: I think the same happens when using `SaveToFileAndDispose()` in the Editor. The textures are not downloaded.

**Reply by atteneder** [reply]
There is support for texture export (embed in the glb), but only for certain shaders (Lit; see #385). Which ones are you using and does it work in the editor?

**Comment by Tripl3Design** [comment]
It is a standard shader. The basemap is a jpg wich is on the server (to
reduce build size).

It doesn't work in the editor either.


Op wo 31 aug. 2022 12:54 schreef Andreas Atteneder ***@***.***

**Reply by atteneder** [reply]
That should work though. Please raise an issue and provide a sample asset that reproduces the error. Thanks

**Comment by PantherStudio1** [comment]
Hello, I'm using glTFast plugin 4.5.0, Unity 2020.3.20f1, I'm facing an issue saying : 'GameObjectExporter' does not contain definition for 
'SaveToStreamAndDispose'. 

If the 'SaveToStreamAndDispose' is removed, then in which version I would be able to use it? 

Please help me with this issue as I'm working on a project where is it needed.
![image](https://user-images.githubusercontent.com/69793193/215509831-f07301e6-2907-4063-9264-0feb08fb448d.png)

The 'GameObjectExport' class has only 'SaveToFileAndDispose' defined in it.
![image](https://user-images.githubusercontent.com/69793193/215510893-2858a629-7846-48a7-a709-8c4e09566775.png)

**Reply by atteneder** [reply]
@PantherStudio1 Export GLBs to a Stream was added in 4.6.0 on 2022-02-23. Please update to a newer version.

**Reply by PantherStudio1** [reply]
Thank you for a quick reply.

**Comment by anksji** [comment]
how to solve this texture loss problem when exporting the model at run time in unity webgl. 

I used this code to export the model async void ModelExport(GameObject[] models)
    {

        // CollectingLogger lets you programatically go through
        // errors and warnings the export raised
        var logger = new CollectingLogger();

        // ExportSettings and GameObjectExportSettings allow you to configure the export
        // Check their respective source for details

        // ExportSettings provides generic export settings
        var exportSettings = new ExportSettings
        {
            Format = GltfFormat.Binary,
            FileConflictResolution = FileConflictResolution.Overwrite,
            // Export everything except cameras or animation
            ComponentMask = ~(ComponentType.Camera | ComponentType.Animation),
            // Boost light intensities 
            LightIntensityFactor = 100f,
        };

        // GameObjectExportSettings provides settings specific to a GameObject/Component based hierarchy
        var gameObjectExportSettings = new GameObjectExportSettings
        {
            // Include inactive GameObjects in export
            OnlyActiveInHierarchy = false,
            // Also export disabled components
            DisabledComponents = true,
            // Only export GameObjects on certain layers
            LayerMask = LayerMask.GetMask("Default", "MyCustomLayer"),
        };

        // GameObjectExport lets you create glTFs from GameObject hierarchies
        var export = new GameObjectExport(exportSettings, gameObjectExportSettings, logger: logger);

        // Example of gathering GameObjects to be exported (recursively)
        var rootLevelNodes = models;

        // Add a scene
        export.AddScene(rootLevelNodes, "My new glTF scene");


        string ProjectPath = Path.GetFullPath(Path.Combine(Application.dataPath, ".."));
        //string WebExportPath = Path.GetFullPath(Path.Combine(ProjectPath, "../../web/UnityLibrary/exports"));
        string WebExportPath = "E:\\myfolder\\model_Export";
        //System.IO.File.WriteAllText(System.Environment.GetFolderPath(System.Environment.SpecialFolder.Desktop) + "\\App\\mytextfile.txt", "Test");

        string fileName = "NewModel_" + DateTime.Now.ToString("yyyyMMdd_HHmmss") + "_" + PublicMethods.GenerateRandomString(3) + ".glb";
        //string fileName = "Aug_model"+PublicMethods.GenerateRandomString(3)+"model"+ PublicMethods.GenerateRandomString(4)+".gltf";
        string filePath = Path.GetFullPath("E:\\myfolder\\model_Export\\" + fileName);
        //string filePath = Path.Combine(WebExportPath, fileName);
        // Async glTF export
        messageText.SetText("Export starts latest... ");
        //saving model in byte
        var outStream = new MemoryStream();
        //saving file in bytes
        //var success = await export.SaveToStreamAndDispose(outStream);
        //saving model in file
        var success = await export.SaveToFileAndDispose(filePath);

        if (!success)
        {
            messageText.SetText("Something went wrong exporting a glTF ");
            Debug.LogError("Something went wrong exporting a glTF");
            // Log all exporter messages
            logger.LogAll();
        }
        else
        {

            messageText.SetText("glTF model export success at location " + filePath);
            StreamToBytes(outStream);
            Debug.Log("glTF model export success at location " + filePath);
        }
    }

**Reply by atteneder** [reply]
Though your problem might be related to one that @Tripl3Design had, I think this should be a regular bug report via an issue.
Also, you need to provide much more details about your problem.
